In [1]:
# Import Dependencies
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

### PRE-PROCESSING DATA

In [2]:
food_access_raw_df = pd.read_csv("https://gtbootcamp20230221.s3.amazonaws.com/FoodAccessResearchAtlasData2019.csv", dtype={'CensusTract': str})
pd.set_option('display.max_rows', None)
food_access_raw_df.head()

CensusTract    State          County  Urban  Pop2010  OHU2010  \
0  01001020100  Alabama  Autauga County      1     1912      693   
1  01001020200  Alabama  Autauga County      1     2170      743   
2  01001020300  Alabama  Autauga County      1     3373     1256   
3  01001020400  Alabama  Autauga County      1     4386     1722   
4  01001020500  Alabama  Autauga County      1    10766     4082   

   GroupQuartersFlag  NUMGQTRS  PCTGQTRS  LILATracts_1And10  ...  \
0                  0       0.0      0.00                  0  ...   
1                  0     181.0      8.34                  1  ...   
2                  0       0.0      0.00                  0  ...   
3                  0       0.0      0.00                  0  ...   
4                  0     181.0      1.68                  0  ...   

   TractSeniors  TractWhite  TractBlack  TractAsian  TractNHOPI  TractAIAN  \
0         221.0      1622.0       217.0        14.0         0.0       14.0   
1         214.0       888.0      1217.0         5.0         0.0        5.0   
2         439.0      2576.0       647.0        17.0         5.0       11.0   
3         904.0      4086.0       193.0        18.0         4.0       11.0   
4        1126.0      8666.0      1437.0       296.0         9.0       48.0   

   TractOMultir  TractHispanic  TractHUNV  TractSNAP  
0          45.0           44.0        6.0      102.0  
1          55.0           75.0       89.0      156.0  
2         117.0           87.0       99.0      172.0  
3          74.0           85.0       21.0       98.0  
4         310.0          355.0      230.0      339.0  

[5 rows x 147 columns]

In [3]:
# Create a subset dataframe with selected columns
food_access_raw_df = food_access_raw_df[["CensusTract", "State", "County", "Urban", "Pop2010", "OHU2010", "PovertyRate",
                                    "MedianFamilyIncome", "LAhalfand10", "LAPOP1_10", "LAPOP05_10", "LAPOP1_20", "LALOWI1_10",
                                     "LALOWI05_10", "LALOWI1_20", "lapophalfshare", "lalowihalfshare", "lakidshalfshare",
                                     "laseniorshalfshare", "lawhitehalfshare", "lablackhalfshare", "laasianhalfshare",
                                     "lanhopihalfshare", "laaianhalfshare", "laomultirhalfshare", "lahisphalfshare", "lahunvhalfshare", "lasnaphalfshare", "lapop10share", "lalowi10share", 
                                     "lakids10share", "laseniors10share", "lawhite10share", "lablack10share",
                                     "laasian10share", "lanhopi10share", "laaian10share", "laomultir10share", "lahisp10share", "lahunv10share",
                                     "lasnap10share", "TractLOWI", "TractKids", "TractSeniors", "TractWhite",
                                     "TractBlack", "TractAsian", "TractNHOPI", "TractAIAN", "TractOMultir",
                                     "TractHispanic", "TractHUNV", "TractSNAP"]]
food_access_raw_df.tail()

CensusTract    State           County  Urban  Pop2010  OHU2010  \
72526  56043000200  Wyoming  Washakie County      0     3326     1317   
72527  56043000301  Wyoming  Washakie County      1     2665     1154   
72528  56043000302  Wyoming  Washakie County      1     2542     1021   
72529  56045951100  Wyoming    Weston County      0     3314     1322   
72530  56045951300  Wyoming    Weston County      1     3894     1699   

       PovertyRate  MedianFamilyIncome  LAhalfand10  LAPOP1_10  ...  \
72526          9.7             67254.0            1      902.0  ...   
72527         11.6             64152.0            1        NaN  ...   
72528         16.3             69605.0            1      155.0  ...   
72529         17.5             74500.0            1      840.0  ...   
72530         17.3             76838.0            1     1376.0  ...   

       TractSeniors  TractWhite  TractBlack  TractAsian  TractNHOPI  \
72526         593.0      3106.0         6.0        15.0         0.0   
72527         399.0      2377.0         5.0        23.0         0.0   
72528         516.0      2312.0        11.0        10.0         1.0   
72529         499.0      3179.0        15.0        10.0         1.0   
72530         650.0      3706.0         6.0        10.0         2.0   

       TractAIAN  TractOMultir  TractHispanic  TractHUNV  TractSNAP  
72526       27.0         172.0          309.0       61.0       64.0  
72527       40.0         220.0          446.0       88.0       41.0  
72528       26.0         182.0          407.0       23.0       64.0  
72529       47.0          62.0           91.0       47.0       34.0  
72530       44.0         126.0          125.0       34.0      110.0  

[5 rows x 53 columns]

In [5]:
# Calculate percentage population of each group
food_access_raw_df[["TractLOWI_PCT", "TractKids_PCT", "TractSeniors_PCT", "TractWhite_PCT",
                "TractBlack_PCT", "TractAsian_PCT", "TractNHOPI_PCT", "TractAIAN_PCT",
                "TractOMultir_PCT", "TractHispanic_PCT"]] = food_access_raw_df[["TractLOWI", "TractKids", "TractSeniors", "TractWhite",
                                     "TractBlack", "TractAsian", "TractNHOPI", "TractAIAN", "TractOMultir",
                                     "TractHispanic"]]\
    .apply(lambda x: x/food_access_raw_df['Pop2010']*100).round(2)


# Display Dataframe
food_access_raw_df.head()

CensusTract    State          County  Urban  Pop2010  OHU2010  PovertyRate  \
0  01001020100  Alabama  Autauga County      1     1912      693         11.3   
1  01001020200  Alabama  Autauga County      1     2170      743         17.9   
2  01001020300  Alabama  Autauga County      1     3373     1256         15.0   
3  01001020400  Alabama  Autauga County      1     4386     1722          2.8   
4  01001020500  Alabama  Autauga County      1    10766     4082         15.2   

   MedianFamilyIncome  LAhalfand10  LAPOP1_10  ...  TractLOWI_PCT  \
0             81250.0            1     1896.0  ...          23.80   
1             49000.0            1     1261.0  ...          36.96   
2             62609.0            1     1552.0  ...          38.72   
3             70607.0            1     1363.0  ...          21.02   
4             96334.0            1     2643.0  ...          20.82   

   TractKids_PCT  TractSeniors_PCT  TractWhite_PCT  TractBlack_PCT  \
0          26.52             11.56           84.83           11.35   
1          27.93              9.86           40.92           56.08   
2          26.50             13.02           76.37           19.18   
3          23.14             20.61           93.16            4.40   
4          29.37             10.46           80.49           13.35   

   TractAsian_PCT  TractNHOPI_PCT  TractAIAN_PCT  TractOMultir_PCT  \
0            0.73            0.00           0.73              2.35   
1            0.23            0.00           0.23              2.53   
2            0.50            0.15           0.33              3.47   
3            0.41            0.09           0.25              1.69   
4            2.75            0.08           0.45              2.88   

   TractHispanic_PCT  
0               2.30  
1               3.46  
2               2.58  
3               1.94  
4               3.30  

[5 rows x 63 columns]

In [6]:
# Calculate percentage TractHUNV and TractSNAP  group
food_access_raw_df[["TractHUNV_PCT", "TractSNAP_PCT"]] = food_access_raw_df[["TractHUNV", "TractSNAP"]]\
    .apply(lambda x: x/food_access_raw_df['OHU2010']*100).round(2)


# Display Dataframe
food_access_raw_df.head()

CensusTract    State          County  Urban  Pop2010  OHU2010  PovertyRate  \
0  01001020100  Alabama  Autauga County      1     1912      693         11.3   
1  01001020200  Alabama  Autauga County      1     2170      743         17.9   
2  01001020300  Alabama  Autauga County      1     3373     1256         15.0   
3  01001020400  Alabama  Autauga County      1     4386     1722          2.8   
4  01001020500  Alabama  Autauga County      1    10766     4082         15.2   

   MedianFamilyIncome  LAhalfand10  LAPOP1_10  ...  TractSeniors_PCT  \
0             81250.0            1     1896.0  ...             11.56   
1             49000.0            1     1261.0  ...              9.86   
2             62609.0            1     1552.0  ...             13.02   
3             70607.0            1     1363.0  ...             20.61   
4             96334.0            1     2643.0  ...             10.46   

   TractWhite_PCT  TractBlack_PCT  TractAsian_PCT  TractNHOPI_PCT  \
0           84.83           11.35            0.73            0.00   
1           40.92           56.08            0.23            0.00   
2           76.37           19.18            0.50            0.15   
3           93.16            4.40            0.41            0.09   
4           80.49           13.35            2.75            0.08   

   TractAIAN_PCT  TractOMultir_PCT  TractHispanic_PCT  TractHUNV_PCT  \
0           0.73              2.35               2.30           0.87   
1           0.23              2.53               3.46          11.98   
2           0.33              3.47               2.58           7.88   
3           0.25              1.69               1.94           1.22   
4           0.45              2.88               3.30           5.63   

   TractSNAP_PCT  
0          14.72  
1          21.00  
2          13.69  
3           5.69  
4           8.30  

[5 rows x 65 columns]

In [26]:
# Create a subset dataframe with selected columns
first_food_access_df = food_access_raw_df[["Pop2010", "OHU2010", "PovertyRate", "MedianFamilyIncome", "TractLOWI", 
                                    "LAhalfand10", "lapophalfshare", "lalowihalfshare", "lakidshalfshare", 
                                    "laseniorshalfshare", "lawhitehalfshare", "lablackhalfshare", 
                                    "laasianhalfshare", "lanhopihalfshare", "laaianhalfshare", "laomultirhalfshare",
                                    "lahisphalfshare", "lahunvhalfshare", "lasnaphalfshare", "TractLOWI_PCT", 
                                    "TractKids_PCT", "TractSeniors_PCT", "TractWhite_PCT", "TractBlack_PCT", 
                                    "TractAsian_PCT", "TractNHOPI_PCT", "TractAIAN_PCT", "TractOMultir_PCT", 
                                    "TractHispanic_PCT", "TractHUNV_PCT", "TractSNAP_PCT"]]
first_food_access_df.tail()

Pop2010  OHU2010  PovertyRate  MedianFamilyIncome  TractLOWI  \
72526     3326     1317          9.7             67254.0      765.0   
72527     2665     1154         11.6             64152.0      774.0   
72528     2542     1021         16.3             69605.0      789.0   
72529     3314     1322         17.5             74500.0      955.0   
72530     3894     1699         17.3             76838.0     1095.0   

       LAhalfand10  lapophalfshare  lalowihalfshare  lakidshalfshare  \
72526            1           99.99            25.36            26.58   
72527            1           27.89            10.49             7.03   
72528            1           35.10            10.25             8.45   
72529            1           81.74            25.26            14.80   
72530            1           64.93            17.24            15.93   

       laseniorshalfshare  ...  TractSeniors_PCT  TractWhite_PCT  \
72526               17.83  ...             17.83           93.39   
72527                4.57  ...             14.97           89.19   
72528                6.80  ...             20.30           90.95   
72529               12.15  ...             15.06           95.93   
72530               10.80  ...             16.69           95.17   

       TractBlack_PCT  TractAsian_PCT  TractNHOPI_PCT  TractAIAN_PCT  \
72526            0.18            0.45            0.00           0.81   
72527            0.19            0.86            0.00           1.50   
72528            0.43            0.39            0.04           1.02   
72529            0.45            0.30            0.03           1.42   
72530            0.15            0.26            0.05           1.13   

       TractOMultir_PCT  TractHispanic_PCT  TractHUNV_PCT  TractSNAP_PCT  
72526              5.17               9.29           4.63           4.86  
72527              8.26              16.74           7.63           3.55  
72528              7.16              16.01           2.25           6.27  
72529              1.87               2.75           3.56           2.57  
72530              3.24               3.21           2.00           6.47  

[5 rows x 31 columns]

In [27]:
# Drop rows with NAN values
first_food_access_df = first_food_access_df.dropna()
 
# Reset the index
first_food_access_df = first_food_access_df.reset_index(drop = True)

first_food_access_df.tail()

Pop2010  OHU2010  PovertyRate  MedianFamilyIncome  TractLOWI  \
67281     3326     1317          9.7             67254.0      765.0   
67282     2665     1154         11.6             64152.0      774.0   
67283     2542     1021         16.3             69605.0      789.0   
67284     3314     1322         17.5             74500.0      955.0   
67285     3894     1699         17.3             76838.0     1095.0   

       LAhalfand10  lapophalfshare  lalowihalfshare  lakidshalfshare  \
67281            1           99.99            25.36            26.58   
67282            1           27.89            10.49             7.03   
67283            1           35.10            10.25             8.45   
67284            1           81.74            25.26            14.80   
67285            1           64.93            17.24            15.93   

       laseniorshalfshare  ...  TractSeniors_PCT  TractWhite_PCT  \
67281               17.83  ...             17.83           93.39   
67282                4.57  ...             14.97           89.19   
67283                6.80  ...             20.30           90.95   
67284               12.15  ...             15.06           95.93   
67285               10.80  ...             16.69           95.17   

       TractBlack_PCT  TractAsian_PCT  TractNHOPI_PCT  TractAIAN_PCT  \
67281            0.18            0.45            0.00           0.81   
67282            0.19            0.86            0.00           1.50   
67283            0.43            0.39            0.04           1.02   
67284            0.45            0.30            0.03           1.42   
67285            0.15            0.26            0.05           1.13   

       TractOMultir_PCT  TractHispanic_PCT  TractHUNV_PCT  TractSNAP_PCT  
67281              5.17               9.29           4.63           4.86  
67282              8.26              16.74           7.63           3.55  
67283              7.16              16.01           2.25           6.27  
67284              1.87               2.75           3.56           2.57  
67285              3.24               3.21           2.00           6.47  

[5 rows x 31 columns]

In [28]:
# Get dataframe information
first_food_access_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67286 entries, 0 to 67285
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Pop2010             67286 non-null  int64  
 1   OHU2010             67286 non-null  int64  
 2   PovertyRate         67286 non-null  float64
 3   MedianFamilyIncome  67286 non-null  float64
 4   TractLOWI           67286 non-null  float64
 5   LAhalfand10         67286 non-null  int64  
 6   lapophalfshare      67286 non-null  float64
 7   lalowihalfshare     67286 non-null  float64
 8   lakidshalfshare     67286 non-null  float64
 9   laseniorshalfshare  67286 non-null  float64
 10  lawhitehalfshare    67286 non-null  float64
 11  lablackhalfshare    67286 non-null  float64
 12  laasianhalfshare    67286 non-null  float64
 13  lanhopihalfshare    67286 non-null  float64
 14  laaianhalfshare     67286 non-null  float64
 15  laomultirhalfshare  67286 non-null  float64
 16  lahi

In [29]:
first_food_access_df['LAhalfand10'].value_counts()

1    49047
0    18239
Name: LAhalfand10, dtype: int64

NOTE: The value counts here are binary; previously, the initial model had 17 features, and the target (‘LAhalfand10’) was 1 = 49047 (low access) and 0 = 22735 (not low access). With this final iteration of the model, I increased the training data to include all the 'la...halfshare' data. While this was highly amenable to increasing the accuracy, it also imbalanced the target somewhat.

### SPLIT AND NORMALIZE

In [30]:
# Split our preprocessed data into our features and target arrays
y = first_food_access_df['LAhalfand10']
X = first_food_access_df.drop(['LAhalfand10'], axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [31]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

### TRAIN

In [32]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 =  8
hidden_nodes_layer2 = 12

nn_model = tf.keras.models.Sequential()

# First hidden layer
nn_model.add(tf.keras.layers.Dense(hidden_nodes_layer1, activation="relu", input_dim= 30))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(hidden_nodes_layer2, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 8)                 248       
                                                                 
 dense_7 (Dense)             (None, 12)                108       
                                                                 
 dense_8 (Dense)             (None, 1)                 13        
                                                                 
Total params: 369
Trainable params: 369
Non-trainable params: 0
_________________________________________________________________


In [33]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [34]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, epochs=80)

Epoch 1/80
1577/1577 [==============================] - 3s 2ms/step - loss: 0.4780 - accuracy: 0.7689
Epoch 2/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3944 - accuracy: 0.8169
Epoch 3/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3784 - accuracy: 0.8231
Epoch 4/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3703 - accuracy: 0.8263
Epoch 5/80
1577/1577 [==============================] - 3s 2ms/step - loss: 0.3643 - accuracy: 0.8284
Epoch 6/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3599 - accuracy: 0.8308
Epoch 7/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3564 - accuracy: 0.8313
Epoch 8/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3539 - accuracy: 0.8331
Epoch 9/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.3519 - accuracy: 0.8335
Epoch 10/80
1577/1577 [==============================] - 2s 2ms/step - loss: 0.350

1577/1577 [==============================] - 2s 2ms/step - loss: 0.3247 - accuracy: 0.8479


### RESULTS

In [35]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

526/526 - 1s - loss: 0.3283 - accuracy: 0.8462 - 747ms/epoch - 1ms/step
Loss: 0.32829371094703674, Accuracy: 0.8461538553237915


In this case, a loss of 0.3283 suggests that the model is making relatively accurate predictions, while an accuracy of 0.8462 suggests that the model is correct in its predictions for approximately 85% of the samples in the dataset. Further optimization such as adding hyperparameter tuning could potentially improve its generalizability, efficiency, or performance on new datasets or tasks. However, the time it takes to optimize this model (greater than 2 hours in earlier iterations) versus the meager yield shown in the previous attempts suggests that these results are both good and should be retained. 